### ISSUE TITLE
SearchField and New Item Producer

### ISSUE THREAD

**Issue by larochef** [dlsc-software-consulting-gmbh/GemsFX#59]
Hello,

I am willing to use the search field as a suggestion field, this means that I would like to let the user create new items when the one he wants doesn't exist. 
I have given a try to the the new item producer property, but then, it is called everytime the user sends a new letter, so if I want to create a new item called cat, then it will try to create "c", "ca" and, at last "cat", I haven't found a way to figure out when the user has finished typing.

How is this feature supposed  to work ?

**Comment by dlemmermann** [comment]
The producer gets invoked every single time to create a new object on-the-fly. But the producer should not add the newly created item to your data, yet. This is up to your application to decide. You might wanna do it when the field looses its focus or when the user hits enter inside the textfield (`searchField.getEditor().setOnAction(eat -> {}`).

**Comment by larochef** [comment]
I see,  thanks for the explanations. I did try to use the getEditor().setOnAction(...) but it didn't work. I guess it is because the enter key event is consumed by the event filter added in the constructor.

**Comment by dlemmermann** [comment]
Then maybe override the` commit()` method (but make sure to call `super.commit()`).

**Comment by larochef** [comment]
I will give it a try, thanks!

**Comment by dlemmermann** [comment]
I wonder what your UI looks like ... do you automatically run some code when the search field contains a selected item? Why exactly is it important to your app to know when the user is done? When I wrote the control I simply wanted to make sure that eventually there is an item that is selected. For me this is normally done inside a dialog and the app continues its execution when the user closes the dialog. When doing so the app will use the selected item.

**Comment by larochef** [comment]
good question, let me try to explain it.
I have some interface where I can tag things. I have a predefined list of tags, but that list must be easy to extend, thus creating tags depending on what the user has typed. Once a tag is selected then the user is presented the next item to tag. 

Since there could be a lot of things to tag, I want to limit at most the actions needed by the user. Today, I have a text field  allowing to create a new tag, or select an existing tag if the same text is typed and some quick buttons next to it to select tags. I'd like to continue being able in the same field to create or select a tag, and have a quicker way to find  existing labels.

Knowing when an element is selected isn't too hard with the component, with the provided property. There is still a first point that the first suggestion is auto-selected, since the component wants to ensure something is selected. Then if the user wants to create something new the looks like an existing suggestion, the menu will keep showing. for instance there could be a "not a cat" tag, and later on, the user will want to create a "cat" tag.

To summarize, the flow I would want is: the user starts to type, it shows things that could match. If he selects one, then use it and move on to next item. If he presses enter, then create a tag with what was entered, use it and move on to next item.

**Comment by larochef** [comment]
I played again with the component, overriding the commit method, and I could get it to work as I wanted.

The trick I used to have the create work is to add a suggestion in the list saying something like `add new item: cat`, with this, no need for more hacks.

**Comment by dlemmermann** [comment]
Glad to hear it ... although I would  not consider overriding the commit() a hack. As you can see in the rest of the code pretty much everything is "final". Having the commit() method public and non-final indicates that it is totally OK for you to override it.
